File: go/beer-song/beer.go
package beer

import (
	"strconv"
	"errors"
	//"fmt"
)

func Song() string{
	song, _ := Verses(99, 0)
	return song
}

func Verses(start, stop int) (string, error){
	if start < stop { return "", errors.New("start should be >= stop") }

	res := ""

	for i := start ; i >= stop ; i-- {
		v, e := Verse(i)

		if e != nil { return "", e }

		res += v
		res += "\n"
	}

	return res, nil
}

func Verse(v int) (string, error) {
	//lazy I know, but tbh i don't like the song exercises.
	if v < 0 || v > 100 {
		return "", errors.New("Verse out of range")
	}else if v == 0 {
		return "No more bottles of beer on the wall, no more bottles of beer.\n" +
			"Go to the store and buy some more, 99 bottles of beer on the wall.\n", nil
	}else if v == 1 {
		return "1 bottle of beer on the wall, 1 bottle of beer.\n" +
			"Take it down and pass it around, no more bottles of beer on the wall.\n", nil
	}else if v == 2 {
		return "2 bottles of beer on the wall, 2 bottles of beer.\n" +
			"Take one down and pass it around, 1 bottle of beer on the wall.\n", nil
	}

	bottles := strconv.Itoa(v) + " bottles of beer"
	res := bottles
	res += " on the wall, "
	res += bottles
	res += ".\n"

	bottles = strconv.Itoa(v-1) + " bottles of beer"
	res += "Take one down and pass it around, "
	res += bottles
	res += " on the wall.\n"

	return res, nil
}

File: haskell/beer-song/src/Beer.hs
module Beer (song) where

import Data.Char (toUpper)

line :: Int -> String 
line x = line1 ++ line2 ++ "\n"
    where
        line1 = sentence1 x
        line2 = sentence2 x

bottles :: Int -> String
bottles x
   | x == -1   = bottles 99
   | x == 0    = "no more bottles of beer"
   | x == 1    = show x ++ " bottle of beer"
   | otherwise = show x ++ " bottles of beer"

sentence1 :: Int -> String
sentence1 x = [toUpper $ head b] ++ (tail b) ++ " on the wall, " ++ b ++ ".\n"
    where 
        b = bottles x       

sentence2 :: Int -> String
sentence2 0 = "Go to the store and buy some more, " ++ bottles (0-1) ++ " on the wall."
sentence2 1 = "Take it down and pass it around, " ++ bottles 0 ++ " on the wall.\n"
sentence2 x = "Take one down and pass it around, " ++ bottles (x-1) ++ " on the wall.\n"

song :: String
song = concatMap line [99,98..0]